versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java

```diff
@@ -0,0 +1,415 @@
+  }
+
+  @Override
+  protected int entitySize(CommitLogEntry entry) {
+    return toProto(entry).getSerializedSize();
+  }
+
+  @Override
+  protected int entitySize(KeyWithType entry) {
+    return toProto(entry).getSerializedSize();
+  }
+
+  @Override
+  protected Stream<KeyListEntity> fetchKeyLists(
+      NonTransactionalOperationContext ctx, List<Hash> keyListsIds) {
+    return fetchMappedPage(
+        client.getKeyLists(),
+        keyListsIds,
+        document -> {
+          Hash hash = idAsHash(document);
+          KeyList keyList = protoToKeyList(data(document));
+          return KeyListEntity.of(hash, keyList);
+        })
+        .stream();
+  }
+
+  @Override
+  protected void writeIndividualCommit(NonTransactionalOperationContext ctx, CommitLogEntry entry)
+      throws ReferenceConflictException {
+    insert(client.getCommitLog(), entry.getHash(), toProto(entry).toByteArray());
+  }
+
+  @Override
+  protected void writeMultipleCommits(
+      NonTransactionalOperationContext ctx, List<CommitLogEntry> entries)
+      throws ReferenceConflictException {
+    List<Document> docs =
+        entries.stream()
+            .map(e -> toDoc(e.getHash(), toProto(e).toByteArray()))
+            .collect(Collectors.toList());
+    insert(client.getCommitLog(), docs);
+  }
+
+  @Override
+  protected void writeKeyListEntities(
+      NonTransactionalOperationContext ctx, List<KeyListEntity> newKeyListEntities) {
+    for (KeyListEntity keyList : newKeyListEntities) {
+      try {
+        insert(client.getKeyLists(), keyList.getId(), toProto(keyList.getKeys()).toByteArray());
+      } catch (ReferenceConflictException e) {
+        throw new IllegalStateException(e);
+      }
+    }
+  }
+
+  @Override
+  protected void writeGlobalCommit(NonTransactionalOperationContext ctx, GlobalStateLogEntry entry)
+      throws ReferenceConflictException {
+    String id = toId(Hash.of(entry.getId()));
+    insert(client.getGlobalLog(), toDoc(id, entry.toByteArray()));
+  }
+
+  @Override
+  protected void unsafeWriteGlobalPointer(
+      NonTransactionalOperationContext ctx, GlobalStatePointer pointer) {
+    Document doc = toDoc(pointer);
+
+    UpdateResult result =
+        client
+            .getGlobalPointers()
+            .updateOne(
+                Filters.eq((Object) globalPointerKey),
+                new Document("$set", doc),
+                new UpdateOptions().upsert(true));
+
+    if (!result.wasAcknowledged()) {
+      throw new IllegalStateException(
+          "Unacknowledged write to " + client.getGlobalPointers().getNamespace());
+    }
+  }
+
+  @Override
+  protected boolean globalPointerCas(
+      NonTransactionalOperationContext ctx,
+      GlobalStatePointer expected,
+      GlobalStatePointer newPointer) {
+    Document doc = toDoc(newPointer);
+    byte[] expectedGlobalId = expected.getGlobalId().toByteArray();
+
+    UpdateResult result =
+        client
+            .getGlobalPointers()
+            .updateOne(
```

Comment:
How has this been tested? From my reading of the [docs](https://docs.mongodb.com/manual/reference/method/db.collection.updateOne/) the `updateOne` method in this way is the best/correct operator. However most google searches about conditional updates talk about pipeline operators which makes me wonder if its possible that the filter in the first argument returns an entry then the 2nd argument modifies it but these two operations aren't done atomically and there could be a race under high load.

Reply:
Mongo docs mention that changes are atomic at a specific document level... but I'll double check.

Reply:
https://docs.mongodb.com/manual/core/write-operations-atomicity/

All write operations are supposed to be atomic at the level of individual documents. I have to assume that `updateOne` is considered one update operation (both the filter part and the change part), so if concurrent operations target the same doc, I suppose one of them will not "find" the changed data, which we catch in the result check (a few lines below).

Having reviewed the docs once more, I believe it might be preferable to use `replaceOne` instead, actually.

Reply:
Switched to `replaceOne` in 681efd77e4b9657f67632fc35c82476ff7283ab1.

I do not think it makes a material difference in term of atomicity / consistency, but it looks cleaner code-wise.

Reply:
Re: "conditional update". From my POV those pipelines are mostly used for deriving the updated value on the server side. We do not need to do that in our case as we have the complete new doc available at the start of `globalPointerCas()`